jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/PathResponseListener.java

```diff
@@ -0,0 +1,196 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jetty.client.Response.Listener;
+import org.eclipse.jetty.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link Response.ContentListener} that produces an {@link Path}
+ * that allows applications to save a file from a response {@link Response}
+ * like curl &lt;URL&gt; -o file.bin does.
+ * <p>
+ * Typical usage is:
+ * <pre>
+ *  httpClient.newRequest(host, port)
+ * .send(new PathResponseListener(Path.of("/tmp/file.bin"));
+ * 
+ *  var request = httpClient.newRequest(host, port);
+ *  CompletableFuture&gt;Path&gt; completable = PathResponseListener.write(request, Path.of("/tmp/file.bin"), rewriteExistingFile);
+ * </pre>
+ */
+public class PathResponseListener extends CompletableFuture<Response> implements Listener
+{
+    private static final Logger LOG = LoggerFactory.getLogger(InputStreamResponseListener.class);
+    
+    private Path path;
+    private Throwable failure;
+    private FileChannel fileOut;
+    private int bytesWrite;
+    
+    public PathResponseListener(Path path, boolean overwrite) throws FileNotFoundException, IOException, FileAlreadyExistsException
+    {           
+        this.path = path;
+        
+        // Throws the exception if file can't be overwritten 
+        // otherwise truncate it.
+        if (this.path.toFile().exists() && !overwrite)
+        {
+            throw new FileAlreadyExistsException("File can't be overwritten");
+        }
+        
+        try
+        {
+            fileOut = FileChannel.open(this.path, StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);
+        }
+        catch (IOException e) 
+        {   
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to instantiate object", e);
+            
+            throw e;
+        }
+    }
+    
+    @Override
+    public void onHeaders(Response response)
+    {
+        if (response.getStatus() != HttpStatus.OK_200)
+        {
+            this.cancel(true);
+            throw new HttpResponseException(String.format("HTTP status code of this response %d", response.getStatus()), response);
+        }
+    }
+    
+    @Override
+    public void onContent(Response response, ByteBuffer content)
+    {
+        try
+        {
+            this.bytesWrite += this.write(content).get();
+            if (LOG.isDebugEnabled())
+                LOG.debug("%d bytes written", bytesWrite);
+        }
+        catch (InterruptedException | ExecutionException e)
+        {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void onComplete(Result result)
+    {
+        if (result.isFailed() && this.failure == null)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Result failure", failure);
+            
+            this.cancel(true);
```

Comment:
You should call `completeExceptionally(result.getFailure())` here.

Reply:
I'm wondering how it will affect dependent `CompletableFutures`? Of course, we don't have other ones now, but in the future. Or is it better to just keep it simple?

Reply:
Not sure what you mean here.
`cancel()` is not a method of `CompletableFuture` -- it is a method of `Future` that should not be used for `CompletableFuture`, because we want to rely on `complete()/completeExceptionally()`.